### PR TITLE
Migrate IMDB to datapipes

### PR DESCRIPTION
**Pull request by kevinchn**
Migrate IMDB dataset to use torchdata datapipes

Reference Issue: #1494

Test:
`pytest test/data/test_builtin_datasets.py -k IMDB`

### CODE REVIEW

File: torchtext/datasets/imdb.py [hunk 1]
```diff
@@ -23,16 +28,25 @@
...
+        hash_dict={os.path.join(root, _PATH): MD5}, hash_type="md5"
+    )
+    cache_dp = HttpReader(cache_dp).end_caching(mode="wb", same_filepath_fn=True)
+    cache_dp = FileOpener(cache_dp, mode="b")
+
+    extracted_files = cache_dp.read_from_tar()
+
+    def filter_imdb_data(key, fname):
+        *_, split, label, file = Path(fname).parts
+        return key == split and (label in ['pos', 'neg'])
```
Comment (Nayef211) [inline torchtext/datasets/imdb.py:48]: In the current implementation, we construct the `['pos', 'neg']` list for every file contained within `extracted_files` datapipe. 

Instead, can we create a `labels` set outside the `filter_imdb_data()` function and then do a contains check to see if the label exists within the `labels` set? This would make it so the set is only created once regardless of the number of files within `extracted_files`. Also, `contains` operations are faster on a set than on a list in Python.

File: torchtext/datasets/imdb.py [hunk 2]
```diff
@@ -23,16 +28,25 @@
...
+
+    filter_extracted_files = extracted_files.filter(lambda t: filter_imdb_data(split, t[0]))
+
+    return filter_extracted_files.readlines(decode=True).map(lambda t: (Path(t[0]).parts[-2], t[1]))
```
Comment (Nayef211) [inline torchtext/datasets/imdb.py:52]: Can we also add double caching similar to what we're doing in the [YelpFullReview dataset PR](https://github.com/pytorch/text/pull/1529/files/1ff59b781c5ee72529f742382e0ad3d0b7871582..61cad246874ba2ac7f6e03f8ab93838f89287540#diff-8056dddf0ce233418c7df18c6c8e9d71e5f81498c0c5b2fe9dcfb793144e1be8R43-R54)?

**Review by Nayef211 — COMMENTED** [review]
Thanks for helping with the dataset migration @kevinchn! Just left a couple of suggestions but otherwise this looks great!

File: torchtext/datasets/imdb.py (hunk 1 above)
Comment (erip) [reply torchtext/datasets/imdb.py:48]: Sets are asymptotically faster with `__contains__`, but with a len-2 iterable, they're about the same. Interestingly, I suspected checking `(label == "pos") or (label == "neg")` would be faster, but it's slower on average...

```python
In [1]: def f1(label):
   ...:     return label in ["pos", "neg"]

In [2]: def f2(label):
   ...:     return label in {"pos", "neg"}

In [3]: def f3(label):
   ...:     return (label == "pos") or (label == "neg")

In [4]: %timeit f1("pos")
90.9 ns ± 1.99 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)

In [5]: %timeit f1("neg")
109 ns ± 1.85 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)

In [6]: %timeit f2("pos")
96 ns ± 2.38 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)

In [7]: %timeit f2("neg")
94.2 ns ± 2.55 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)

In [8]: %timeit f3("pos")
100 ns ± 3.85 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)

In [9]: %timeit f3("neg")
120 ns ± 2.84 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)
```

**Comment by kevinchn** [comment]
Updated. Please take a look again!

File: torchtext/datasets/imdb.py (hunk 1 above)
Comment (Nayef211) [reply torchtext/datasets/imdb.py]: Thanks @erip for helping verify this! You're right that there is a negligible difference in performance when the length of the iterable is small. I just thought it would be better practice but I guess this would be more of a nit comment 😃

File: torchtext/datasets/imdb.py [hunk 3]
```diff
@@ -23,16 +28,35 @@
...
+    url_dp = IterableWrapper([URL])
+
+    cache_compressed_dp = url_dp.on_disk_cache(
+        filepath_fn=lambda x: os.path.join(root, _PATH),
+        hash_dict={os.path.join(root, _PATH): MD5}, hash_type="md5"
+    )
+    cache_compressed_dp = HttpReader(cache_compressed_dp).end_caching(mode="wb", same_filepath_fn=True)
+
+    labels = {"neg", "pos"}
+    cache_decompressed_dp = cache_compressed_dp.on_disk_cache(filepath_fn=lambda x: [os.path.join(root, "aclImdb", split, label) for label in labels])
```
Comment (Nayef211) [inline torchtext/datasets/imdb.py:44]: nit: can we change `"aclImdb"` to `"aclImdb_v1"`. We have been trying to keep the names of the uncompressed folder and the tar archive the same for our datasets.

File: torchtext/datasets/imdb.py [hunk 4]
```diff
@@ -23,16 +28,35 @@
...
+        # eg. fname = "aclImdb/train/neg/12416_3.txt"
+        *_, split, label, file = Path(fname).parts
+        return key == split and label in labels
+
+    cache_decompressed_dp = cache_decompressed_dp.filter(lambda t: filter_imdb_data(split, t[0]))
+
+    cache_decompressed_dp = cache_decompressed_dp.map(lambda t: (Path(t[0]).parts[-2], t[1]))  # eg. "aclImdb/train/neg/12416_3.txt" -> "neg"
+    cache_decompressed_dp = cache_decompressed_dp.readlines(decode=True)
+    cache_decompressed_dp = cache_decompressed_dp.lines_to_paragraphs()  # group by label in cache file
+    cache_decompressed_dp = cache_decompressed_dp.end_caching(mode="wt", filepath_fn=lambda x: os.path.join(root, "aclImdb", split, x))
```
Comment (Nayef211) [inline torchtext/datasets/imdb.py:58]: Same comment as above

File: torchtext/datasets/imdb.py [hunk 5]
```diff
@@ -23,16 +28,35 @@
...
+    cache_decompressed_dp = cache_decompressed_dp.read_from_tar()
+
+    def filter_imdb_data(key, fname):
+        # eg. fname = "aclImdb/train/neg/12416_3.txt"
+        *_, split, label, file = Path(fname).parts
+        return key == split and label in labels
+
+    cache_decompressed_dp = cache_decompressed_dp.filter(lambda t: filter_imdb_data(split, t[0]))
+
+    cache_decompressed_dp = cache_decompressed_dp.map(lambda t: (Path(t[0]).parts[-2], t[1]))  # eg. "aclImdb/train/neg/12416_3.txt" -> "neg"
```
Comment (Nayef211) [inline torchtext/datasets/imdb.py:55]: nit: update comment to say `"aclImdb_v1"`

File: torchtext/datasets/imdb.py [hunk 6]
```diff
@@ -23,16 +28,35 @@
...
+    cache_decompressed_dp = cache_decompressed_dp.readlines(decode=True)
+    cache_decompressed_dp = cache_decompressed_dp.lines_to_paragraphs()  # group by label in cache file
+    cache_decompressed_dp = cache_decompressed_dp.end_caching(mode="wt", filepath_fn=lambda x: os.path.join(root, "aclImdb", split, x))
+
+    data_dp = FileOpener(cache_decompressed_dp, mode="t")
+    # get label from cache file, eg. "aclImdb/train/neg" -> "neg"
```
Comment (Nayef211) [inline torchtext/datasets/imdb.py:61]: nit: update comment to say `"aclImdb_v1"`

**Review by Nayef211 — APPROVED** [review]
This LGTM. Let's merge this in after fixing the nit comments!

File: torchtext/datasets/imdb.py (hunk 5 above)
Comment (kevinchn) [reply torchtext/datasets/imdb.py:55]: For this one, `aclImdb` refers to the actual folder name inside the tar. I'll keep it as is if there's no objections

Comment (Nayef211) [reply torchtext/datasets/imdb.py]: Gotcha that makes sense
